src/tree/blanked_tree.rs

```diff
@@ -0,0 +1,180 @@
+            Ok(f(node, U::default()))
+        } else {
+            let parent = self.parent(*node_index)?;
+            let parent_result = self.direct_path_map(&parent, f)?;
+            Ok(f(self.node_mut(node_index)?, parent_result))
+        }
+    }
+
+    /// Compute a function f based on the node itself, as well as the result of
+    /// the same function computed on the left and right child. Leaves return
+    /// the result of the function with their node, as well as the default
+    /// values for `U`. Returns an error if the given index is not within the
+    /// tree.
+    pub(crate) fn fold_tree<F, U: Default>(
+        &self,
+        node_index: &NodeIndex,
+        f: &F,
+    ) -> Result<U, BinaryTreeError>
+    where
+        F: Fn(&NodeIndex, &U, &U) -> U,
+    {
+        if node_index.is_leaf() {
+            Ok(f(node_index, &U::default(), &U::default()))
+        } else {
+            let left_node = self.left(*node_index)?;
+            let left_result = self.fold_tree(&left_node, f)?;
+            let right_node = self.right(*node_index)?;
+            let right_result = self.fold_tree(&right_node, f)?;
+            Ok(f(node_index, &left_result, &right_result))
+        }
+    }
+
+    /// Returns a vector with leaf indices, where each index corresponds to a
+    /// blank leaf. The vector is ordered by the index of the leaf.
+    fn free_leaves(&self) -> Vec<LeafIndex> {
+        let mut free_leaves = Vec::new();
+        for index in 0..self.leaf_count().as_usize() {
+            let leaf_index = LeafIndex::from(index);
+            // We can unwrap here, because index is scoped to be within the
+            // tree.
+            if self.leaf(&leaf_index).unwrap().is_none() {
+                free_leaves.push(leaf_index);
+            }
+        }
+        free_leaves
+    }
+
+    /// Remove fully-blank subtrees on the right side.
+    pub(crate) fn trim(&mut self) {
+        let mut right_most_index = self.size().as_usize() - 1;
+        // We can unwrap here, because the right-most index is always within the
+        // tree.
+        while self.is_blank(&NodeIndex::from(right_most_index)).unwrap()
+            && self
+                .is_blank(&NodeIndex::from(right_most_index - 1))
+                .unwrap()
+            && self.size() > NodeIndex::from(2 as usize)
+        {
+            // We can unwrap here, because we know that the outtermost nodes are
+            // blank and the tree is large enough to remove two nodes.
+            self.remove(2).unwrap();
+            right_most_index -= 2;
+        }
+    }
+
+    /// Add a number of leaves into the first blank leaves. If not enough blank
+    /// leaves exist, the tree is extended using blanks as intermediate nodes.
+    /// Finally, the tree is trimmed after the operation. The returned vector
+    /// contains the indices of the newly added leaves.
+    pub(crate) fn add_blanked(&mut self, mut new_nodes: Vec<T>) -> Vec<NodeIndex> {
```

Comment:
Maybe `add_leaves` is a better name?

I think the function should return `Vec<LeafIndex>`, we should use type safety wherever we can.

Reply:
Good ideas! I agree and I changed it that way.